### ISSUE TITLE
Memory usage a bit high?

### ISSUE THREAD

**Issue by ngocphamm** [lwouis/alt-tab-macos#285]
Hi, it's me again...

Please feel free to ignore/close this if this is totally unreasonable. I just noticed memory usage of the app. From my perspective, It think the app is mostly background, and probably should have used less memory. But I really don't know what's the "standard" for memory usage here, so just want to let you know my thought. Or maybe there's an uncovered bug with memory leak ¯\\\_(ツ)_/¯

![2020-04-28 at 4 38 PM](https://user-images.githubusercontent.com/395123/80535487-dddede80-896e-11ea-9edd-e251985b63c8.png)

**Comment by lwouis** [comment]
Hey @ngocphamm!

I'm not sure where we stand on that front. ~100MB seems reasonable to me, but it wouldn't hurt to take some time to analyze. From my limited experience doing C++ work, finding memory leaks in these low-level languages can be really nasty. At least finding what's taking up the 100MB should be easy.

I suspect one of the culprit is the pool of views i used to reduce the latency of showing the UI because we allocate a bunch of views upfront instead of just-in-time when the shortcut is pressed

**Comment by ngocphamm** [comment]
Yeah defintely not end of the world (well except for COVID-19) yet, so, again I appreciate your willing to look into this!

**Comment by lwouis** [comment]
To my surprise, by far most of the memory is coming from the OS calls to get the windows screenshots. It's responsible for the large majority of allocations and memory usage:

![image](https://user-images.githubusercontent.com/106195/80559799-b729b200-8a19-11ea-83ec-f523859a0ea9.png)

The app itself is using very little memory:

![image](https://user-images.githubusercontent.com/106195/80559945-26070b00-8a1a-11ea-9eb4-498978213420.png)

It's all about the screenshots:

![image](https://user-images.githubusercontent.com/106195/80559970-40d97f80-8a1a-11ea-8a17-6dee5c797ef3.png)

[Opened a ticket to the Preferences pod](https://github.com/sindresorhus/Preferences/issues/49), as it seems they may be leaking a few hundred bytes at init.

**Comment by lwouis** [comment]
`CGImage`s coming from the the OS APIs are indeed massive. On my macbook pro retina display, a fullscreen window represents:

* 3104 × 1978 pixels
* **~20MB for the `CGImage` object** (uncompressed image)
* 209 KB for the JPG encoding of that image

So if you have 20 windows open, that's 20 * 20 = 400 MB of data coming from the OS API. It can way worse with more open windows, and more importantly, with a bigger screen like an 5k or 8k display.

**Comment by ngocphamm** [comment]
That sounds like a problem maybe if one has many many windows across mutiple screens. I don't notice any delay that would annoy me yet, but I imagine that can slow the app down to display the thumbnails for some people 🤔

The explanation makes total sense for me, as I'm using 2 external (Full-HD only) monitors with my laptop, so more screen estate to make images of, thus the memory usage. 

Thanks for digging in to this @lwouis!

**Comment by lwouis** [comment]
Actually I discovered that some constants I used as parameters to the private API that gives the fast screenshots of the windows, those parameters were actually not correct.

On retina display, each pixel generated by the OS UIs is spread on 4px. Currently, we got the full 3104 × 1978 image which is useless resolution because it's just stretching the pixels. Making a thumbnail for AltTab has the same exact quality by starting from the internal resolution which is 4 times smaller.

I'll include that change as well as the tiny memory leaks in the next release. That should help HiDPI screen users :)
